src/agents/coordinator_agent.py

```diff
@@ -0,0 +1,172 @@
+from messages.modify_reservation import ModifyReservation
+from messages.parking_availability import ParkingAvailable
+from messages.reservation_request import RequestReservation
+from messages.response import ReservationResponse
+
+
+class RegionalCoordinator(Agent):
+    def __init__(self, jid, password, x_from, x_to, y_from, y_to, parking_agents_jids):
+        super().__init__(jid, password)
+        self._x_min = x_from
+        self._x_max = x_to
+        self._y_min = y_from
+        self._y_max = y_to
+        self._parking_agents_jids = parking_agents_jids  # list of parking agents jids within the region
+        # {user_jid: {parkings: [parking_available, ...], destination: (x, y)}}
+        self._per_user_data = {}
+
+    def _prepare_check_offers_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "query-ref")
+        template.set_metadata("action", "check-offers")
+        return template
+
+    def _prepare_make_reservation_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "request")
+        template.set_metadata("action", "make-reservation")
+        return template
+
+    def _prepare_modify_reservation_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "request")
+        template.set_metadata("action", "modify-reservation")
+        return template
+
+    def _prepare_parking_availability_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "inform")
+        template.set_metadata("action", "parking-availability")
+        return template
+
+    def _prepare_reservation_response_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "inform")
+        template.set_metadata("action", "reservation-response")
+        return template
+
+    def _calculate_distance(self, x_user, y_user, x_parking, y_parking):
+        return ((x_user - x_parking) ** 2 + (y_user - y_parking) ** 2) ** 0.5
+
+    def _consolidate_offers_per_user(self, user) -> ConsolidatedOffers:
+        dest_x, dest_y = self._per_user_data[user]["destination"]
+        free_parkings = [parking for parking in self._per_user_data[user]["parkings"] if parking.available]
+        offers = []
+        for parking in free_parkings:
+            offer = Offer(parking_id=parking.parking_id, price=parking.parking_price,
+                          distance=self._calculate_distance(dest_x, dest_y, parking.parking_x, parking.parking_y))
+            offers.append(offer)
+        offers = sorted(offers, key=lambda x: x.distance)
+        offers = offers[:3]
```

Comment:
How about creating a param / an argument to decide the number of offers to be returned?

Reply:
Hmm, as we wrote in documentation, RegionalCoordinator should return only 3 best offers. Maybe I will move it to constants?